test/uk/gov/hmrc/traderservices/controllers/FormFieldMappingsSpec.scala

```diff
@@ -16,34 +16,150 @@
-      validateName("@Artur") shouldBe invalid
-      validateName("Ar#tur") shouldBe invalid
-      validateName("ĄĘÓŚŻĆŁąęółśćńżźāēīūčģķļņšž") shouldBe Valid
+
+    "validate EPU" in {
+      epuMapping.bind(Map("" -> "123")) shouldBe Right(EPU("123"))
+      epuMapping.bind(Map("" -> "")) should haveOnlyError[EPU]("error.epu.required")
+      epuMapping.bind(Map("" -> "1")) should haveOnlyError[EPU]("error.epu.invalid-length")
+      epuMapping.bind(Map("" -> "12")) should haveOnlyError[EPU]("error.epu.invalid-length")
+      epuMapping.bind(Map("" -> "1224")) should haveOnlyError[EPU]("error.epu.invalid-length")
+      epuMapping.bind(Map("" -> "12A")) should haveOnlyError[EPU]("error.epu.invalid-only-digits")
+      epuMapping.bind(Map("" -> "AAA")) should haveOnlyError[EPU]("error.epu.invalid-only-digits")
+      epuMapping.bind(Map("" -> "A12")) should haveOnlyError[EPU]("error.epu.invalid-only-digits")
+      epuMapping.bind(Map("" -> "1A2")) should haveOnlyError[EPU]("error.epu.invalid-only-digits")
+      epuMapping.bind(Map("" -> "701")) should haveOnlyError[EPU]("error.epu.invalid-number")
+      epuMapping.bind(Map("" -> "999")) should haveOnlyError[EPU]("error.epu.invalid-number")
+    }
+
+    "validate EntryNumber" in {
+      entryNumberMapping.bind(Map("" -> "000000Z")) shouldBe Right(EntryNumber("000000Z"))
+      entryNumberMapping.bind(Map("" -> "A00000Z")) shouldBe Right(EntryNumber("A00000Z"))
+      entryNumberMapping.bind(Map("" -> "")) should haveOnlyError[EntryNumber]("error.entryNumber.required")
+      entryNumberMapping.bind(Map("" -> "00000Z")) should haveOnlyError[EntryNumber]("error.entryNumber.invalid-length")
+      entryNumberMapping.bind(Map("" -> "0000Z")) should haveOnlyError[EntryNumber]("error.entryNumber.invalid-length")
+      entryNumberMapping.bind(Map("" -> "00000")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-length",
+        "error.entryNumber.invalid-ends-with-letter"
+      )
+      entryNumberMapping.bind(Map("" -> "0")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-length",
+        "error.entryNumber.invalid-ends-with-letter"
+      )
+      entryNumberMapping.bind(Map("" -> "Z")) should haveOnlyError[EntryNumber]("error.entryNumber.invalid-length")
+      entryNumberMapping.bind(Map("" -> "+")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-length",
+        "error.entryNumber.invalid-only-digits-and-letters",
+        "error.entryNumber.invalid-ends-with-letter"
+      )
+      entryNumberMapping.bind(Map("" -> "000000Z+")) should haveOnlyErrors[EntryNumber](
+        "error.entryNumber.invalid-only-digits-and-letters",
+        "error.entryNumber.invalid-length",
+        "error.entryNumber.invalid-ends-with-letter",
+        "error.entryNumber.invalid-letter-wrong-position"
+      )
+      entryNumberMapping.bind(Map("" -> "000+000Z")) should haveOnlyErrors[EntryNumber](
```

Comment:
Technically I think the Z is in the right position here (at the end) & it follows a number so I'm not sure if we need to show this message? "error.entryNumber.invalid-letter-wrong-position" (Unless it's correct to have because it's in the wrong position in terms of being the 8th character which is invalid?)

Reply:
I will try to correct this

Reply:
I'm still a little unsure on this error. (Same thing on this page line 48-51 & on ConsignmentDetailsFormSpec.scala lines 72-77). 

In these three instances the Z is at the end of the Entry Number which is correct positioning, and even though they are incorrect lengthwise I think if we were to show it this error might be confusing to users if they have got the letter at the end? 

(Happy to discuss this with you in case there's something I'm not quite getting)